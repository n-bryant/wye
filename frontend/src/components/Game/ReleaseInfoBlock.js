import React from "react";
import PropTypes from "prop-types";

import { withStyles } from "@material-ui/core/styles";
import { withWidth } from "@material-ui/core/";
import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../../lib/classNamePrefix";

import Link from "next/link";

import Icon from "@mdi/react";
import {
  mdiWindows,
  mdiApple,
  mdiLinux,
  mdiSteam,
  mdiGoogleController,
  mdiGoogleControllerOff
} from "@mdi/js";

import Grid from "@material-ui/core/Grid";
import Typography from "@material-ui/core/Typography";
import Tooltip from "@material-ui/core/Tooltip";

import styles from "./ReleaseInfoBlock.styles";

export const PLATFORM_MAP = {
  windows: mdiWindows,
  mac: mdiApple,
  linux: mdiLinux,
  default: mdiSteam
};

/**
 * renders an info block for a game containing release details
 */
export const ReleaseInfoBlock = props => {
  const classnames = ReleaseInfoBlock.classnames(props);
  const { mdWidth, data, width } = props;

  // the short description is provided as an html string
  const createMarkup = () => {
    return { __html: data.shortDescription };
  };

  // renders a header image for the info block
  const HeaderImg = () => {
    const [loaded, setLoaded] = React.useState(false);

    const onLoad = () => {
      setLoaded(true);
    };

    return (
      <div className={classnames.element("headerImgContainer")}>
        <img
          className={classnames.element("headerImg", {
            hidden: !loaded
          })}
          src={data.headerImageUrl}
          alt={data.name}
          onLoad={onLoad}
        />
      </div>
    );
  };

  return (
    <Grid
      className={classnames.root({
        withAdjustedPadding: ["xs", "sm"].some(val => val === width)
      })}
      container
      item
      xs={12}
      md={mdWidth}
      spacing={1}
    >
      <Grid item xs={12}>
        <HeaderImg />
      </Grid>
      <Grid container item xs={12}>
        <div dangerouslySetInnerHTML={createMarkup()} />
        <div className={classnames.element("infoContainer")}>
          <Typography variant="body2">
            <span className={classnames.element("infoType")}>
              Release Date:{" "}
            </span>
            {data.releaseDate}
          </Typography>
          <Typography variant="body2">
            <span className={classnames.element("infoType")}>Publishers: </span>
            {data.publishers.join(", ")}
          </Typography>
          <Typography variant="body2">
            <span className={classnames.element("infoType")}>Developers: </span>
            {data.developers.join(", ")}
          </Typography>
          <Typography
            className={classnames.element("iconContainer")}
            variant="body2"
          >
            <span className={classnames.element("infoType")}>Platforms: </span>
            {data.platforms.map((platform, index) => (
              <Tooltip key={index} title={platform}>
                <Icon
                  className={classnames.element("icon")}
                  path={
                    PLATFORM_MAP[platform]
                      ? PLATFORM_MAP[platform]
                      : PLATFORM_MAP.default
                  }
                />
              </Tooltip>
            ))}
          </Typography>
          <Typography
            className={classnames.element("iconContainer")}
            variant="body2"
          >
            <span className={classnames.element("infoType")}>
              Controller Support:{" "}
            </span>
            <Tooltip title={data.controllerSupport.toString()}>
              <Icon
                className={classnames.element("icon")}
                path={
                  data.controllerSupport
                    ? mdiGoogleController
                    : mdiGoogleControllerOff
                }
              />
            </Tooltip>
          </Typography>
          <Typography variant="body2">
            <span className={classnames.element("infoType")}>Website: </span>
            <Link href={data.website} prefetch={false}>
              <a className={classnames.element("link")} target="_blank">
                {data.website}
              </a>
            </Link>
          </Typography>
        </div>
      </Grid>
    </Grid>
  );
};
ReleaseInfoBlock.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}ReleaseInfoBlock`
);
ReleaseInfoBlock.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    withAdjustedPadding: PropTypes.string,
    headerImgContainer: PropTypes.string,
    headerImg: PropTypes.string,
    headerImgHidden: PropTypes.string,
    infoContainer: PropTypes.string,
    infoType: PropTypes.string,
    iconContainer: PropTypes.string,
    link: PropTypes.string,
    icon: PropTypes.string
  }),
  // the allowed grid width at md+ screen sizes
  mdWidth: PropTypes.number,
  // data for the game
  data: PropTypes.shape({
    name: PropTypes.string,
    headerImageUrl: PropTypes.string,
    shortDescription: PropTypes.string,
    platforms: PropTypes.array,
    controllerSupport: PropTypes.bool,
    developers: PropTypes.array,
    publishers: PropTypes.array,
    releaseDate: PropTypes.string,
    website: PropTypes.string
  }),
  // width from material-ui breakpoints
  width: PropTypes.string
};
ReleaseInfoBlock.defaultProps = {
  classes: {},
  mdWidth: 5
};

export default withWidth()(withStyles(styles)(ReleaseInfoBlock));
