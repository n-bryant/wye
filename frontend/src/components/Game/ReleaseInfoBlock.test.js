import Icon from "@mdi/react";
import { mdiGoogleController, mdiGoogleControllerOff } from "@mdi/js";
import { ReleaseInfoBlock } from "./ReleaseInfoBlock";

describe("ReleaseInfoBlock", () => {
  const props = {
    classes: {
      root: "root",
      withAdjustedPadding: "withAdjustedPadding",
      headerImgContainer: "headerImgContainer",
      headerImg: "headerImg",
      headerImgHidden: "headerImgHidden",
      infoContainer: "infoContainer",
      infoType: "infoType",
      iconContainer: "iconContainer",
      link: "link",
      icon: "icon"
    },
    width: "sm",
    mdWidth: 5,
    data: {
      name: "name",
      headerImageUrl: "headerImageUrl",
      shortDescription: "shortDescription",
      platforms: ["windows", "mac", "linux", "fake"],
      controllerSupport: "full",
      developers: ["developer1", "developer2"],
      publishers: ["publishers"],
      releaseDate: "releaseDate",
      website: "website"
    }
  };

  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<ReleaseInfoBlock {...props} />);
  });

  it("should render successfully", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should adjust the padding if the width prop value is 'xs' or 'sm'", () => {
    // sm/xs
    expect(
      wrapper.findWhere(n => n.hasClass(props.classes.withAdjustedPadding))
        .length
    ).toBe(1);
    wrapper.setProps({
      width: "xs"
    });
    wrapper.update();
    expect(
      wrapper.findWhere(n => n.hasClass(props.classes.withAdjustedPadding))
        .length
    ).toBe(1);

    // md
    wrapper.setProps({
      width: "md"
    });
    wrapper.update();
    expect(
      wrapper.findWhere(n => n.hasClass(props.classes.withAdjustedPadding))
        .length
    ).toBe(0);
  });

  it("should set the controller icon based on the controllerSupport value of the data prop", () => {
    // "full"
    expect(
      wrapper
        .find(Icon)
        .at(4)
        .prop("path")
    ).toBe(mdiGoogleController);

    // "partial"
    wrapper.setProps({
      data: {
        ...props.data,
        controllerSupport: "partial"
      }
    });
    wrapper.update();
    expect(
      wrapper
        .find(Icon)
        .at(4)
        .prop("path")
    ).toBe(mdiGoogleController);

    // undefined
    wrapper.setProps({
      data: {
        ...props.data,
        controllerSupport: undefined
      }
    });
    wrapper.update();
    expect(
      wrapper
        .find(Icon)
        .at(4)
        .prop("path")
    ).toBe(mdiGoogleControllerOff);
  });
});
