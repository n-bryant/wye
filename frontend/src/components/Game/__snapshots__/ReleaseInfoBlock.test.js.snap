// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReleaseInfoBlock should render successfully 1`] = `
<WithStyles(ForwardRef(Grid))
  className="wye-ReleaseInfoBlock wye-ReleaseInfoBlock--withAdjustedPadding root withAdjustedPadding"
  container={true}
  item={true}
  md={5}
  spacing={1}
  xs={12}
>
  <WithStyles(ForwardRef(Grid))
    item={true}
    xs={12}
  >
    <HeaderImg />
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    container={true}
    item={true}
    xs={12}
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "shortDescription",
        }
      }
    />
    <div
      className="wye-ReleaseInfoBlock__infoContainer infoContainer"
    >
      <WithStyles(ForwardRef(Typography))
        variant="body2"
      >
        <span
          className="wye-ReleaseInfoBlock__infoType infoType"
        >
          Release Date:
           
        </span>
        releaseDate
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        variant="body2"
      >
        <span
          className="wye-ReleaseInfoBlock__infoType infoType"
        >
          Publishers: 
        </span>
        publishers
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        variant="body2"
      >
        <span
          className="wye-ReleaseInfoBlock__infoType infoType"
        >
          Developers: 
        </span>
        developer1, developer2
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        className="wye-ReleaseInfoBlock__iconContainer iconContainer"
        variant="body2"
      >
        <span
          className="wye-ReleaseInfoBlock__infoType infoType"
        >
          Platforms: 
        </span>
        <WithStyles(ForwardRef(Tooltip))
          key="0"
          title="windows"
        >
          <Icon
            className="wye-ReleaseInfoBlock__icon icon"
            color={null}
            horizontal={false}
            path="M3,12V6.75L9,5.43V11.91L3,12M20,3V11.75L10,11.9V5.21L20,3M3,13L9,13.09V19.9L3,18.75V13M20,13.25V22L10,20.09V13.1L20,13.25Z"
            rotate={0}
            size={null}
            spin={false}
            vertical={false}
          />
        </WithStyles(ForwardRef(Tooltip))>
        <WithStyles(ForwardRef(Tooltip))
          key="1"
          title="mac"
        >
          <Icon
            className="wye-ReleaseInfoBlock__icon icon"
            color={null}
            horizontal={false}
            path="M18.71,19.5C17.88,20.74 17,21.95 15.66,21.97C14.32,22 13.89,21.18 12.37,21.18C10.84,21.18 10.37,21.95 9.1,22C7.79,22.05 6.8,20.68 5.96,19.47C4.25,17 2.94,12.45 4.7,9.39C5.57,7.87 7.13,6.91 8.82,6.88C10.1,6.86 11.32,7.75 12.11,7.75C12.89,7.75 14.37,6.68 15.92,6.84C16.57,6.87 18.39,7.1 19.56,8.82C19.47,8.88 17.39,10.1 17.41,12.63C17.44,15.65 20.06,16.66 20.09,16.67C20.06,16.74 19.67,18.11 18.71,19.5M13,3.5C13.73,2.67 14.94,2.04 15.94,2C16.07,3.17 15.6,4.35 14.9,5.19C14.21,6.04 13.07,6.7 11.95,6.61C11.8,5.46 12.36,4.26 13,3.5Z"
            rotate={0}
            size={null}
            spin={false}
            vertical={false}
          />
        </WithStyles(ForwardRef(Tooltip))>
        <WithStyles(ForwardRef(Tooltip))
          key="2"
          title="linux"
        >
          <Icon
            className="wye-ReleaseInfoBlock__icon icon"
            color={null}
            horizontal={false}
            path="M14.62,8.35C14.2,8.63 12.87,9.39 12.67,9.54C12.28,9.85 11.92,9.83 11.53,9.53C11.33,9.37 10,8.61 9.58,8.34C9.1,8.03 9.13,7.64 9.66,7.42C11.3,6.73 12.94,6.78 14.57,7.45C15.06,7.66 15.08,8.05 14.62,8.35M21.84,15.63C20.91,13.54 19.64,11.64 18,9.97C17.47,9.42 17.14,8.8 16.94,8.09C16.84,7.76 16.77,7.42 16.7,7.08C16.5,6.2 16.41,5.3 16,4.47C15.27,2.89 14,2.07 12.16,2C10.35,2.05 9,2.81 8.21,4.4C8,4.83 7.85,5.28 7.75,5.74C7.58,6.5 7.43,7.29 7.25,8.06C7.1,8.71 6.8,9.27 6.29,9.77C4.68,11.34 3.39,13.14 2.41,15.12C2.27,15.41 2.13,15.7 2.04,16C1.85,16.66 2.33,17.12 3.03,16.96C3.47,16.87 3.91,16.78 4.33,16.65C4.74,16.5 4.9,16.6 5,17C5.65,19.15 7.07,20.66 9.24,21.5C13.36,23.06 18.17,20.84 19.21,16.92C19.28,16.65 19.38,16.55 19.68,16.65C20.14,16.79 20.61,16.89 21.08,17C21.57,17.09 21.93,16.84 22,16.36C22.03,16.1 21.94,15.87 21.84,15.63"
            rotate={0}
            size={null}
            spin={false}
            vertical={false}
          />
        </WithStyles(ForwardRef(Tooltip))>
        <WithStyles(ForwardRef(Tooltip))
          key="3"
          title="fake"
        >
          <Icon
            className="wye-ReleaseInfoBlock__icon icon"
            color={null}
            horizontal={false}
            path="M12,2A10,10 0 0,1 22,12A10,10 0 0,1 12,22C7.4,22 3.55,18.92 2.36,14.73L6.19,16.31C6.45,17.6 7.6,18.58 8.97,18.58C10.53,18.58 11.8,17.31 11.8,15.75V15.62L15.2,13.19H15.28C17.36,13.19 19.05,11.5 19.05,9.42C19.05,7.34 17.36,5.65 15.28,5.65C13.2,5.65 11.5,7.34 11.5,9.42V9.47L9.13,12.93L8.97,12.92C8.38,12.92 7.83,13.1 7.38,13.41L2,11.2C2.43,6.05 6.73,2 12,2M8.28,17.17C9.08,17.5 10,17.13 10.33,16.33C10.66,15.53 10.28,14.62 9.5,14.29L8.22,13.76C8.71,13.58 9.26,13.57 9.78,13.79C10.31,14 10.72,14.41 10.93,14.94C11.15,15.46 11.15,16.04 10.93,16.56C10.5,17.64 9.23,18.16 8.15,17.71C7.65,17.5 7.27,17.12 7.06,16.67L8.28,17.17M17.8,9.42C17.8,10.81 16.67,11.94 15.28,11.94C13.9,11.94 12.77,10.81 12.77,9.42A2.5,2.5 0 0,1 15.28,6.91C16.67,6.91 17.8,8.04 17.8,9.42M13.4,9.42C13.4,10.46 14.24,11.31 15.29,11.31C16.33,11.31 17.17,10.46 17.17,9.42C17.17,8.38 16.33,7.53 15.29,7.53C14.24,7.53 13.4,8.38 13.4,9.42Z"
            rotate={0}
            size={null}
            spin={false}
            vertical={false}
          />
        </WithStyles(ForwardRef(Tooltip))>
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        className="wye-ReleaseInfoBlock__iconContainer iconContainer"
        variant="body2"
      >
        <span
          className="wye-ReleaseInfoBlock__infoType infoType"
        >
          Controller Support:
           
        </span>
        <WithStyles(ForwardRef(Tooltip))
          title="false"
        >
          <Icon
            className="wye-ReleaseInfoBlock__icon icon"
            color={null}
            horizontal={false}
            path="M2,5.27L3.28,4L20,20.72L18.73,22L12.73,16H7.97L5,19C4.67,19.3 4.23,19.5 3.75,19.5A1.75,1.75 0 0,1 2,17.75V17.5L3,10.12C3.1,9.09 3.53,8.17 4.19,7.46L2,5.27M5,10V11H7V13H8V11.27L6.73,10H5M16.5,6C18.86,6 20.79,7.81 21,10.12L22,17.5V17.75C22,18.41 21.64,19 21.1,19.28L7.82,6H16.5M16.5,8A0.75,0.75 0 0,0 15.75,8.75A0.75,0.75 0 0,0 16.5,9.5A0.75,0.75 0 0,0 17.25,8.75A0.75,0.75 0 0,0 16.5,8M14.75,9.75A0.75,0.75 0 0,0 14,10.5A0.75,0.75 0 0,0 14.75,11.25A0.75,0.75 0 0,0 15.5,10.5A0.75,0.75 0 0,0 14.75,9.75M18.25,9.75A0.75,0.75 0 0,0 17.5,10.5A0.75,0.75 0 0,0 18.25,11.25A0.75,0.75 0 0,0 19,10.5A0.75,0.75 0 0,0 18.25,9.75M16.5,11.5A0.75,0.75 0 0,0 15.75,12.25A0.75,0.75 0 0,0 16.5,13A0.75,0.75 0 0,0 17.25,12.25A0.75,0.75 0 0,0 16.5,11.5Z"
            rotate={0}
            size={null}
            spin={false}
            vertical={false}
          />
        </WithStyles(ForwardRef(Tooltip))>
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        variant="body2"
      >
        <span
          className="wye-ReleaseInfoBlock__infoType infoType"
        >
          Website: 
        </span>
        <Link
          href="website"
          prefetch={false}
        >
          <a
            className="wye-ReleaseInfoBlock__link link"
            target="_blank"
          >
            website
          </a>
        </Link>
      </WithStyles(ForwardRef(Typography))>
    </div>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
